dissect/evidence/ewf.py

```diff
@@ -152,83 +154,130 @@ def __init__(self, fh):
 
             segment.offset = segment_offset * self.volume.sector_size
             segment.sector_offset = segment_offset
             segment_offset += segment.sector_count
 
-            self.segments.append(segment)
-
-        if not self.header or not self.volume or not self.segments:
+        if not self.header or not self.volume:
             raise EWFError(f"Failed to load EWF: {fh}")
 
         self.chunk_size = self.volume.sector_count * self.volume.sector_size
 
         max_size = self.volume.chunk_count * self.volume.sector_count * self.volume.sector_size
-        last_table = self.segments[-1].tables[-1]
-        last_chunk_size = len(last_table.read_chunk(last_table.header.num_entries - 1))
+        last_table = self.open_segment(len(self.fh) - 1).tables[-1]
+        last_chunk_size = len(last_table.read_chunk(last_table.num_entries - 1))
 
         self.size = max_size - (self.chunk_size - last_chunk_size)
-        super().__init__(self.size)
 
-    def read_sectors(self, sector, count):
-        log.debug("EWF::read_sectors(0x%x, 0x%x)", sector, count)
-        r = []
+    def open_segment(self, idx: int) -> Segment:
+        # Poor mans LRU
+        if idx in self._segments:
+            self._segment_lru.remove(idx)
+            self._segment_lru.append(idx)
+            return self._segments[idx]
 
-        segment_idx = bisect_right(self.segment_offsets, sector)
-        while count > 0:
-            segment = self.segments[segment_idx]
+        if len(self._segment_lru) >= MAX_OPEN_SEGMENTS:
+            oldest_idx = self._segment_lru.pop(0)
+            oldest_segment = self._segments.pop(oldest_idx)
 
-            segment_remaining_sectors = segment.sector_count - (sector - segment.sector_offset)
-            segment_sectors = min(segment_remaining_sectors, count)
+            # Don't close it if we received it as a file-like object
+            if not hasattr(self.fh[oldest_idx], "read"):
+                oldest_segment.fh.close()
 
-            r.append(segment.read_sectors(sector, segment_sectors))
-            sector += segment_sectors
-            count -= segment_sectors
+            del oldest_segment
 
-            segment_idx += 1
+        fh = self.fh[idx]
+        if not hasattr(fh, "read"):
+            if isinstance(fh, Path):
+                fh = fh.open("rb")
+            else:
+                fh = open(fh, "rb")
 
-        return b"".join(r)
+        segment = Segment(self, fh)
+        if self.volume and 0 < idx <= len(self._segment_offsets):
+            # We already have a known segment offset for this segment, so set it back
+            segment_offset = self._segment_offsets[idx - 1]
+            segment.offset = segment_offset * self.volume.sector_size
+            segment.sector_offset = segment_offset
+        else:
+            # Otherwise we're in the initialization loop (or we're idx == 0)
+            segment.offset = 0
+            segment.sector_offset = 0
```

Comment:
I think there's some value to making the distinction that they are set later. Otherwise the issue we just debugged would've been a lot harder to track down.

Reply:
Done